src/main/java/com/google/sps/data/OrderHandler.java

```diff
@@ -164,4 +164,33 @@ Entity getClosestLibrary(List<Entity> libraries, Point address) throws ApiExcept
 
     return datastore.prepare(orderQuery).asList(FetchOptions.Builder.withDefaults());
   }
+
+  /**
+   * The method called by the servlet that actually places the order in the datastore.
+   */
+
+  public void placeOrder(String uid) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Get the cart entity to get the list of books.
+    Query query = new Query("UserCart").setFilter(new Query.FilterPredicate("uid", Query.FilterOperator.EQUAL, uid));
+    PreparedQuery results = datastore.prepare(query);
+    Entity cartEntity = results.asSingleEntity();
+    // Get the user entity for the address.
+    Query queryUser = new Query("UserInfo")
+        .setFilter(new Query.FilterPredicate("uid", Query.FilterOperator.EQUAL, uid));
+    PreparedQuery resultsUser = datastore.prepare(queryUser);
+    Entity userEntity = resultsUser.asSingleEntity();
+    try {
+      makeOrders(uid, Point.fromAddress((String) userEntity.getProperty("address")),
+          (ArrayList<String>) cartEntity.getProperty("books"));
+    } catch (ApiException | InterruptedException | DataNotFoundException | BadRequestException e) {
+      // Cannot throw exceptions because DoPost only throws IOException.
+      e.printStackTrace();
```

Comment:
Why can't you throw all exceptions in `placeOrder` and catch them in `doPost()`?

Reply:
I will fix this.